freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/rpcserver.py

```diff
@@ -0,0 +1,1258 @@
+        return self.__apps[key]
+
+    def __contains__(self, key):
+        return key in self.__apps
+
+    def __call__(self, environ, start_response):
+        logger.debug('WSGI wsgi_dispatch.__call__:')
+        try:
+            return self.route(environ, start_response)
+        finally:
+            destroy_context()
+
+    def _on_finalize(self):
+        self.url = self.env['mount_ipa']
+        super(wsgi_dispatch, self)._on_finalize()
+
+    def route(self, environ, start_response):
+        key = environ.get('PATH_INFO')
+        if key in self.__apps:
+            app = self.__apps[key]
+            return app(environ, start_response)
+        url = environ['SCRIPT_NAME'] + environ['PATH_INFO']
+        return self.not_found(environ, start_response, url,
+                              'URL fragment "%s" does not have a handler' % (key))
+
+    def mount(self, app, key):
+        """
+        Mount the WSGI application *app* at *key*.
+        """
+#        if self.__islocked__():
+#            raise Exception('%s.mount(): locked, cannot mount %r at %r' % (
+#                self.name, app, key)
+#            )
+        if key in self.__apps:
+            raise Exception('%s.mount(): cannot replace %r with %r at %r' % (
+                self.name, self.__apps[key], app, key)
+            )
+        logger.debug('Mounting %r at %r', app, key)
+        self.__apps[key] = app
+
+
+
+
+
+class WSGIExecutioner(Executioner):
+    """
+    Base class for execution backends with a WSGI application interface.
+    """
+
+    headers = None
+    content_type = None
+    key = ''
+
+    _system_commands = {}
+
+    def _on_finalize(self):
+        self.url = self.env.mount_ipa + self.key
+        super(WSGIExecutioner, self)._on_finalize()
+        if 'wsgi_dispatch' in self.api.Backend:
+            self.api.Backend.wsgi_dispatch.mount(self, self.key)
+
+    def _get_command(self, name):
+        try:
+            # assume version 1 for unversioned command calls
+            command = self.api.Command[name, '1']
+        except KeyError:
+            try:
+                command = self.api.Command[name]
+            except KeyError:
+                command = None
+
+        if command is None or isinstance(command, Local):
+            raise errors.CommandError(name=name)
+
+        return command
+
+    def wsgi_execute(self, environ):
+        result = None
+        error = None
+        _id = None
+        name = None
+        args = ()
+        options = {}
+        command = None
+
+        e = None
+        if not 'HTTP_REFERER' in environ:
+            return self.marshal(result, RefererError(referer='missing'), _id)
+        if not environ['HTTP_REFERER'].startswith('https://%s/ipa' % self.api.env.host) and not self.env.in_tree:
+            return self.marshal(result, RefererError(referer=environ['HTTP_REFERER']), _id)
+        logger.info('X-cdp-request-ID : %s', environ.get('HTTP_X_CDP_REQUEST_ID'))
```

Comment:
Due to licensing reasons, I think we may need to use salt to add the new lines to this file rather than load the entire file.

Reply:
@jamisonbennett @lacikaaa  Instead of a file drop, now I am doing a line write. It will only write one line to the python file to log the trace-id. And I run salt multiple times, it will only write once if the line does not exist. 

Re tested everything and works.